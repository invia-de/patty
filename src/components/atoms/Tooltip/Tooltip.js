import React from 'react';
import PropTypes from 'prop-types';
import styles from './Tooltip.module.scss';
import cx from '../../../utils/classnames';

let globalCounter = 0;

/**
 * @author [Heydon Pickering](https://inclusive-components.design/tooltips-toggletips/)
 * @author [Eric Zieger](mailto:eric.zieger@invia.de)
 */
export default class Tooltip extends React.Component {
  constructor() {
    super();

    this.state = {
      visible: false,
      counter: globalCounter++
    };

    this.handleShow = this.handleShow.bind(this);
    this.handleHide = this.handleHide.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.handleEscape = this.handleEscape.bind(this);
  }

  handleEscape(event) {
    if (event.keyCode === 27 && this.state.visible) {
      this.setState({ visible: false });
    }
  }

  handleShow() {
    if (!this.state.visible) {
      this.setState({
        visible: true
      });
    }
  }

  handleHide() {
    this.setState({
      visible: false,
      click: false
    });
  }

  handleClick(event) {
    if (this.props.onClick) {
      event.persist();
      this.props.onClick(event);
    }

    this.setState({
      visible: true
    });
  }

  render() {
    const {
      message,
      children,
      classNameMessage,
      styleMessage,
      position,
      className,
      style,
      showArrow
    } = this.props;

    return (
      <div
        className={cx(styles.tooltip, className)}
        style={style}
        onMouseEnter={this.handleShow}
        onMouseLeave={this.handleHide}
      >
        <div
          style={styleMessage}
          className={cx(
            this.state.visible
              ? styles['tooltip__message--visible-' + position]
              : styles.tooltip__message,
            showArrow && styles['tooltip__message--arrow-' + position],
            classNameMessage
          )}
          role="tooltip"
          id={`tooltip-${this.state.counter}`}
        >
          {message}
        </div>
        <button
          type="button"
          className={styles.tooltip__button}
          aria-describedby={`tooltip-${this.state.counter}`}
          onFocus={this.handleShow}
          onBlur={this.handleHide}
          onClick={this.handleClick}
          onKeyUp={this.handleEscape}
        >
          {children}
        </button>
      </div>
    );
  }
}
Tooltip.propTypes = {
  /** additional classNames you want to add */
  className: PropTypes.string,
  /** additional styling apart of the className */
  style: PropTypes.object,
  /** required due to accessibility */
  children: PropTypes.node.isRequired,
  /** the tooltip content */
  message: PropTypes.node.isRequired,
  /** optional render position for the tooltip message */
  position: PropTypes.oneOf([
    'top',
    'top-left',
    'top-right',
    'right',
    'bottom',
    'left'
  ]),
  /** optional classNameMessage for the tooltip message */
  classNameMessage: PropTypes.string,
  /** optional style object for the tooltip message */
  styleMessage: PropTypes.object,
  /** callback when the tooltip trigger is clicked */
  onClick: PropTypes.func,
  /** show arrow */
  showArrow: PropTypes.bool
};
Tooltip.defaultProps = {
  position: 'top',
  styleMessage: {}
};
