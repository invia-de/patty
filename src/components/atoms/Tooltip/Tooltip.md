```jsx
<Tooltip message={'message'}>tooltip</Tooltip>
```
