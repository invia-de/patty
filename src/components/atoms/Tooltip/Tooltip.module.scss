@import 'theme.scss';

$tooltip-button_font-size: 18px !default;
$tooltip-message_padding: 10px !default;
$tooltip-message_font-size: 14px !default;
$tooltip-message_box-shadow: 0 1px 4px 0 $blue-light !default;
$tooltip-message_border: 0 !default;
$tooltip-message_background-color: $white !default;
$tooltip-message_arrow-color: $blue-light !default;

.tooltip {
  position: relative;
  display: inline-block;
}

.tooltip__button {
  padding: 2px;
  margin: 0;
  background: transparent;
  border: 0;
  cursor: help;
  color: currentColor;
  text-align: left;
  font-size: $tooltip-button_font-size;
}

.tooltip__message {
  display: none;
}

.tooltip__message--visible-top + .tooltip__button,
.tooltip__message--visible-right + .tooltip__button,
.tooltip__message--visible-left + .tooltip__button,
.tooltip__message--visible-bottom + .tooltip__button {
  position: relative;
  z-index: 9001;
}

.tooltip__message--visible-top {
  font-weight: 400;
  position: absolute;
  bottom: calc(100% + 1rem);
  left: 50%;
  transform: translateX(-50%);
  background: $tooltip-message_background-color;
  font-size: $tooltip-message_font-size;
  padding: $tooltip-message_padding;
  z-index: 9000;
  display: block;
  box-shadow: $tooltip-message_box-shadow;
  border: $tooltip-message_border;
}

.tooltip__message--visible-right {
  composes: tooltip__message--visible-top;
  bottom: auto;
  left: calc(100% + 1rem);
  top: 50%;
  transform: translateY(-50%);
}

.tooltip__message--visible-bottom {
  composes: tooltip__message--visible-top;
  bottom: auto;
  top: calc(100% + 1rem);
}

.tooltip__message--visible-left {
  composes: tooltip__message--visible-top;
  bottom: auto;
  left: auto;
  right: calc(100% + 1rem);
  top: 50%;
  transform: translateY(-50%);
}

.tooltip__message--arrow-top {
  &:after,
  &:before {
    content: '';
    display: block;
    position: absolute;
    width: 0;
    height: 0;
    border-style: solid;
  }

  &:after {
    bottom: -10px;
    left: calc(50% - 10px);
    border-width: 5px 10px;
    border-color: $tooltip-message_background-color transparent transparent;
  }

  &:before {
    bottom: -12px;
    left: calc(50% - 11px);
    border-width: 6px 11px;
    border-color: $tooltip-message_arrow-color transparent transparent;
  }
}

.tooltip__message--arrow-right {
  composes: tooltip__message--arrow-top;

  &:after {
    bottom: calc(50% - 10px);
    left: -10px;
    border-width: 10px 5px;
    border-color: transparent $tooltip-message_background-color transparent
      transparent;
  }

  &:before {
    bottom: calc(50% - 11px);
    left: -12px;
    border-width: 11px 6px;
    border-color: transparent $tooltip-message_arrow-color transparent
      transparent;
  }
}

.tooltip__message--arrow-bottom {
  composes: tooltip__message--arrow-top;

  &:after {
    bottom: 100%;
    left: calc(50% - 10px);
    border-color: transparent transparent $tooltip-message_background-color;
  }

  &:before {
    bottom: 100%;
    left: calc(50% - 11px);
    border-color: transparent transparent $tooltip-message_arrow-color;
  }
}

.tooltip__message--arrow-left {
  composes: tooltip__message--arrow-right;

  &:after {
    left: 100%;
    border-color: transparent transparent transparent
      $tooltip-message_background-color;
  }

  &:before {
    left: 100%;
    border-color: transparent transparent transparent
      $tooltip-message_arrow-color;
  }
}
