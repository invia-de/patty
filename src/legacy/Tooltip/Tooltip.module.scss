@import '../../style/_colors';

.tooltip {
  position: relative;
  display: inline-block;
}

.tooltip__button {
  padding: 2px;
  margin: 0;
  background: transparent;
  border: 0;
  cursor: help;
  color: currentColor;
  text-align: left;
  font-size: 18px;
}

.tooltip__message {
  display: none;
}

.tooltip__message--visible {
  font-weight: 400;
  position: absolute;
  bottom: calc(100% + 0.75rem);
  left: 50%;
  transform: translateX(-50%);
  background: $white;
  font-size: 14px;
  padding: 10px 10px;
  z-index: 9000;
  display: block;
  box-shadow: 0 1px 4px 0 $black-10;
  border: 1px solid $black-10;
}

.tooltip__message--visible + .tooltip__button {
  position: relative;
  z-index: 9001;
}
