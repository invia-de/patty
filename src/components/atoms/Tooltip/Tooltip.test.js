import React from 'react';
import Tooltip from './Tooltip';
import { render } from 'react-testing-library';

test('Tooltip renders', () => {
  const { container } = render(<Tooltip />);
  expect(container.firstChild).toMatchSnapshot();
});
