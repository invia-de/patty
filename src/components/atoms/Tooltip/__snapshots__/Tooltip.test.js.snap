// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip message renders on hover 1`] = `
<div
  class="tooltip"
>
  <div
    class="tooltip__message--visible-top"
    id="tooltip-0"
    role="tooltip"
  >
    tooltip message
  </div>
  <button
    aria-describedby="tooltip-0"
    class="tooltip__button"
    type="button"
  >
    open tooltip
  </button>
</div>
`;
